examples/job-shop-scheduling/jobShopScheduler.py

```diff
@@ -0,0 +1,268 @@
+        key = self.key_function(item)
+        return key
+
+
+class JobShopScheduler:
+    def __init__(self, job_dict, max_time=None):
+        self.tasks = []
+        self.last_task_indices = []
+        self.max_time = max_time
+        self.csp = dwavebinarycsp.ConstraintSatisfactionProblem(dwavebinarycsp.BINARY)
+
+        # Populates self.tasks and self.max_time
+        self._process_data(job_dict)
+
+    def _process_data(self, jobs):
+        """Process user input into a format that is more convenient for JobShopScheduler functions.
+        """
+        # Create and concatenate Task objects
+        tasks = []
+        last_task_indices = [-1]    # -1 for zero-indexing
+        total_time = 0  # total time of all jobs
+
+        for job_name, job_tasks in jobs.items():
+            last_task_indices.append(last_task_indices[-1] + len(job_tasks))
+
+            for i, (machine, time_span) in enumerate(job_tasks):
+                tasks.append(Task(job_name, i, machine, time_span))
+                total_time += time_span
+
+        # Update values
+        self.tasks = tasks
+        self.last_task_indices = last_task_indices[1:]
+
+        if self.max_time is None:
+            self.max_time = total_time
+
+    def _add_one_start_constraint(self):
+        """self.csp gets the constraint: A task can start once and only once
+        """
+        for task in self.tasks:
+            task_times = {get_label(task, t) for t in range(self.max_time)}
+            self.csp.add_constraint(sum_to_one, task_times)
+
+    def _add_precedence_constraint(self):
+        """self.csp gets the constraint: Task must follow a particular order.
+         Note: assumes self.tasks are sorted by jobs and then by position
+        """
+        valid_edges = {(0, 0), (1, 0), (0, 1)}
+        for i, current_task in enumerate(self.tasks[:-1]):
+            next_task = self.tasks[i + 1]
+
+            if current_task.job != next_task.job:
+                continue
+
+            # Forming constraints with the relevant times of the next task
+            for t in range(self.max_time):
+                current_label = get_label(current_task, t)
+
+                for tt in range(min(t + current_task.duration, self.max_time)):
+                    next_label = get_label(next_task, tt)
+                    self.csp.add_constraint(valid_edges, {current_label, next_label})
+
+    def _add_share_machine_constraint(self):
+        """self.csp gets the constraint: At most one task per machine per time
+        """
+        sorted_tasks = sorted(self.tasks, key=lambda x: x.machine)
+        wrapped_tasks = KeyList(sorted_tasks, lambda x: x.machine) # Key wrapper for bisect function
+
+        head = 0
+        valid_values = {(0, 0), (1, 0), (0, 1)}
+        while head < len(sorted_tasks):
+
+            # Find tasks that share a machine
+            tail = bisect_right(wrapped_tasks, sorted_tasks[head].machine)
+            same_machine_tasks = sorted_tasks[head:tail]
```

Comment:
As elegant as this binary search approach is, it seems like a bit of an overkill here.

Namely, once you have your list of tasks sorted, a simple [`itertools.groupby`](https://docs.python.org/3/library/itertools.html#itertools.groupby) is all you need.

(Clarification: I'm not saying you should change your code!)